### PR TITLE
Async pool integration

### PR DESCRIPTION
**Pull request by Eitu33**
https://github.com/massalabs/massa/discussions/1990

### CODE REVIEW

File: massa-execution-worker/Cargo.toml [hunk 1]
```diff
@@ -22,10 +22,11 @@ tracing = "0.1"
 massa_execution_exports = { path = "../massa-execution-exports" }
 massa_models = { path = "../massa-models" }
 massa_hash = { path = "../massa-hash" }
-massa-sc-runtime = { git = "https://github.com/massalabs/massa-sc-runtime", tag = "v0.4.4" }
+massa-sc-runtime = { git = "https://github.com/massalabs/massa-sc-runtime" }
```
Comment (AurelienFT) [inline massa-execution-worker/Cargo.toml:25]: Add a note to retag

**Review by AurelienFT — COMMENTED** [review]
I don't see any Rust problems.

File: massa-execution-worker/src/context.rs [hunk 2]
```diff
@@ -198,7 +212,7 @@ impl ExecutionContext {
     /// * final_state: thread-safe access to the final state. Note that this will be used only for reading, never for writing
     ///
     /// # returns
-    /// A ExecutionContext instance ready for a read-only execution
+    /// A ExecutionContext instance
```
[the remark this answers is not in the document]
Comment (Eitu33) [reply massa-execution-worker/src/context.rs:226]: Copy pasted comment from above this is not what the function does.

File: massa-execution-worker/src/context.rs [hunk 3]
```diff
@@ -355,7 +370,7 @@ impl ExecutionContext {
         self.speculative_ledger.has_data_entry(address, key)
     }
 
-    /// gets the bytecode of an address if it exists in the speculative ledger, or returns None
+    /// note: define this comment
```
[the remark this answers is not in the document]
Comment (Eitu33) [reply massa-execution-worker/src/context.rs:384]: Copy pasted comment from above this is not what the function does, and I wanted to talk about it with someone else.

[the remark this answers is not in the document]
Comment (Eitu33) [reply massa-execution-worker/src/context.rs:384]: Copy pasted comment from above this is not what the function does, and I wanted to talk about it with someone else.

File: massa-execution-worker/src/context.rs [hunk 4]
```diff
@@ -413,4 +428,15 @@ impl ExecutionContext {
         self.speculative_ledger
             .transfer_parallel_coins(from_addr, to_addr, amount)
     }
+
+    // note: needs doc
```
[the remark this answers is not in the document]
Comment (Eitu33) [reply massa-execution-worker/src/context.rs:443]: This is only personal, there shouldn't be any notes when the PR is ready to be merged.

File: massa-async-pool/src/pool.rs [hunk 5]
```diff
@@ -62,16 +62,16 @@ impl AsyncPool {
...
                 // add a new message to the pool
                 (msg_id, AddOrDelete::Add(msg)) => {
-                    self.messages.insert(msg_id, msg);
+                    self.messages.insert(*msg_id, msg.clone());
                 }
 
                 // delete a message from the pool
                 (msg_id, AddOrDelete::Delete) => {
-                    self.messages.remove(&msg_id);
+                    self.messages.remove(msg_id);
```
[the remark this answers is not in the document]
Comment (Eitu33) [reply massa-async-pool/src/pool.rs:74]: That's my bad I updated the function a some point, then reverted without checking thinking this was the previous implementation.

File: massa-execution-worker/src/speculative_async_pool.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        self.new_messages.push(msg);
+    }
+
+    pub fn compute_and_add_changes(&mut self, slot: Slot) -> Vec<AsyncMessage> {
+        let mut pool_copy = self.final_state.read().async_pool.clone();
+        pool_copy.apply_changes_unchecked(std::mem::take(&mut self.previous_changes));
+        for msg in &self.new_messages {
+            // note: message is not pushed if mul overflows
+            // not sure if it is the right way to handle this
+            if let Some(cost) = msg.gas_price.checked_mul_u64(msg.max_gas) {
```
[the remark this answers is not in the document]
Comment (Eitu33) [reply massa-execution-worker/src/speculative_async_pool.rs:61]: This was still a WIP I should've put it in draft, it's on me.

File: massa-node/src/main.rs [hunk 7]
```diff
@@ -126,10 +127,15 @@ async fn launch() -> (
     let ledger_config = LedgerConfig {
         initial_sce_ledger_path: SETTINGS.ledger.initial_sce_ledger_path.clone(),
     };
+    let async_pool_config = AsyncPoolConfig {
+        // note: define this
```
[the remark this answers is not in the document]
Comment (AureliaDolo) [reply massa-node/src/main.rs:131]: It should be read from the config file, as it can be user defined

**Comment by AureliaDolo** [comment]
@Eitu33 I think this pr was closed by bors because the branch it targeted was deleted because the associated pr was merged

**Pull request closed without merging**